### PR TITLE
(maint) Create Base class for User/Group ADSI objects

### PR DESCRIPTION
**Pull request by ekinanp**
There's some shared code in the User/Group ADSI classes. This
commit moves these over to a common Base ADSI class.

### CODE REVIEW

**Comment by ekinanp** [comment]
jenkins please test this on windows10ent-64a,windows10ent-32a,windows2008r2-64a,windows2012r2-64a,windows2012r2_ja-64a%7Blocale=ja%7D,windows2016-64a

File: lib/puppet/util/windows/adsi.rb [hunk 1]
```diff
@@ -106,138 +106,201 @@ def execquery(query)
...
-        # for instance AUTORITE NT on French Windows
-        Puppet::Util::Windows::SID.name_to_principal('SYSTEM').domain.upcase
-      ]
-    end
-
-    def uri(name, host = '.')
-      host = '.' if (localized_domains << Socket.gethostname.upcase).include?(host.upcase)
+  # Common Base class shared by the User and Group
+  # classes below.
+  class Base
```
[the remark this answers is not in the document]
Comment (ekinanp) [reply lib/puppet/util/windows/adsi.rb]: Yes, that is better.

File: lib/puppet/util/windows/adsi.rb [hunk 2]
```diff
@@ -106,138 +106,201 @@ def execquery(query)
...
+      end
 
-      Puppet::Util::Windows::ADSI.uri(name, account_type, host)
-    end
 
-    def parse_name(name)
-      if name =~ /\//
-        raise Puppet::Error.new( _("Value must be in DOMAIN\\user style syntax") )
+      def delete(name)
+        Puppet::Util::Windows::ADSI.delete(name, @object_class)
```
[the remark this answers is not in the document]
Comment (ekinanp) [reply lib/puppet/util/windows/adsi.rb]: Not sure what you mean -- the old code did the same thing: https://github.com/puppetlabs/puppet/blob/master/lib/puppet/util/windows/adsi.rb#L368-L370 and https://github.com/puppetlabs/puppet/blob/master/lib/puppet/util/windows/adsi.rb#L528-L530

[the remark this answers is not in the document]
Comment (caseywilliams) [reply lib/puppet/util/windows/adsi.rb]: Oh, I had matched this to the wrong part of the old file, nevermind

**Comment by ekinanp** [comment]
Updated the PR to change `Base => ADSIObject`.

File: spec/integration/util/windows/adsi_spec.rb [hunk 3]
```diff
@@ -13,7 +13,7 @@
       #     HRESULT error code:0x800708ad
       #           The user name could not be found.
       # Matching on error code alone is sufficient
```
Comment (speedofdark) [inline spec/integration/util/windows/adsi_spec.rb]: But matching on only the error code doesn't match the intent described in lines 11-12 immediately above. Perhaps we need to pull the comments? Or rephrase?

File: lib/puppet/util/windows/adsi.rb [hunk 4]
```diff
@@ -106,138 +106,201 @@ def execquery(query)
...
-      sids = []
-      adsi_child_collection.each do |m|
-        sids << Puppet::Util::Windows::SID.ads_to_principal(m)
+        object = Puppet::Util::Windows::ADSI.connect(uri(*parse_name(name_or_sid)))
+        object.Class.downcase == @object_class
+      rescue
+        # special accounts like SYSTEM or special groups like Authenticated Users cannot
+        # resolve via monikers like WinNT://./SYSTEM,user or WinNT://./Authenticated Users,group
+        # -- they'll fail to connect. thus, given a validly resolved SID, this failure is
+        # ambiguous as it may indicate either a group like Service or an account like SYSTEM
```
Comment (speedofdark) [inline lib/puppet/util/windows/adsi.rb]: Is it ok to be ambiguous?

File: spec/integration/util/windows/adsi_spec.rb (hunk 3 above)
Comment (ekinanp) [reply spec/integration/util/windows/adsi_spec.rb]: Good questions. This was here before the refactor -- https://github.com/puppetlabs/puppet/blob/master/lib/puppet/util/windows/adsi.rb#L209-L224 and https://github.com/puppetlabs/puppet/blob/master/lib/puppet/util/windows/adsi.rb#L439-L454

I don't want to rephrase comments made before this refactor here b/c this PR's really meant to purely move stuff around without changing the original intent.

File: lib/puppet/util/windows/adsi.rb (hunk 4 above)
Comment (ekinanp) [reply lib/puppet/util/windows/adsi.rb]: This was here before the refactor (https://github.com/puppetlabs/puppet/blob/master/lib/puppet/util/windows/adsi.rb#L522-L524), so I think that question's beyond the scope of this PR.

File: lib/puppet/util/windows/adsi.rb [hunk 5]
```diff
@@ -106,138 +106,201 @@ def execquery(query)
...
+      end
+  
+      def uri(name, host = '.')
+        host = '.' if (localized_domains << Socket.gethostname.upcase).include?(host.upcase)
+        Puppet::Util::Windows::ADSI.uri(name, @object_class, host)
+      end
+  
+      def parse_name(name)
+        if name =~ /\//
+          raise Puppet::Error.new( _("Value must be in DOMAIN\\user style syntax") )
```
Comment (Iristyle) [inline lib/puppet/util/windows/adsi.rb]: `DOMAIN\\#{@object_class} style syntax` ?

**Pull request merged**